opds_import.py

```diff
@@ -288,53 +360,148 @@ def import_from_metadata(
+            even_if_not_apparently_updated=True,
+            mirror=self.mirror,
+            http_get=self.http_get,
+        )
+        circulationdata.apply(
+            license_pool, replace=policy
+        )
+
         if license_pool is None:
             # Without a LicensePool, we can't create a Work.
-            self.log.warn(
-                "No LicensePool present for Edition %r, not attempting to create Work.",
-                edition
-            )
-        else:
-            license_pool.presentation_edition = edition
-            work, is_new_work = license_pool.calculate_work(
-                known_edition=edition,
-                even_if_no_author=even_if_no_author,
-            )
-            if work:
-                if immediately_presentation_ready:
-                    # We want this book to be presentation-ready
-                    # immediately upon import. As long as no crucial
-                    # information is missing (like language or title),
-                    # this will do it.
-                    work.set_presentation_ready_based_on_content()
+            error_message = "No LicensePool created for CirculationData %r, will not attempt to continue to create Work." % circulationdata
+            self.log.warn(error_message)
+            raise DBImportException(error_message)
+
+        return license_pool
 
-        return edition
 
-    def extract_metadata(self, feed):
-        """Turn an OPDS feed into a list of Metadata objects and a list of
-        messages.
+    def import_work_from_circulationdata(
+            self, license_pool, even_if_no_author, immediately_presentation_ready
+    ):
+        """ Making work is not based on pool's cutoff logic, but based on whether have a pool and and edition, period.
+        Don't mind if they're new or not.
+        """
+
+        # Note: pool.calculate_work will call self.set_presentation_edition(), 
+        # which will find editions attached to same Identifier.
+        # If the pool has no edition, pool.calculate_work is responsible for not allowing the work to be created.
+        work = None
+        work, is_new_work = license_pool.calculate_work(even_if_no_author=even_if_no_author,)
+        # if pool.calculate_work made a new work, it already called calculate_presentation
+        if (work and not is_new_work):
+            work.calculate_presentation()
+        if (work):
+            if immediately_presentation_ready:
+                # We want this book to be presentation-ready
+                # immediately upon import. As long as no crucial
+                # information is missing (like language or title),
+                # this will do it.
+                work.set_presentation_ready_based_on_content()
+        return license_pool, work
+
+
+    def extract_feed_data(self, feed):
+        """Turn an OPDS feed into lists of Metadata and CirculationData objects, 
+        with associated messages and next_links.
```

Comment:
Right now you have some code that treats CirculationData as a property of Metadata, and some code (such as here) that treats Metadata and CirculationData as totally separate things that need to be kept in separate lists. It's good that CirculationData can be used separately, but given that `Metadata.circulation` is allowed, turning one OPDS feed into a list of Metadata (each of which may or may not have a `.circulation`) yields one list, which is a lot easier to deal with than two.

If you see future refactoring that gets rid of `Metadata.circulation` then it would be appropriate to leave this alone, since we're partway there, but like having `Metadata.circulation` and I think you do as well. What kind of refactoring are you thinking of?